compiler/rustc_errors/src/emitter.rs

```diff
@@ -1676,16 +1676,24 @@ impl EmitterWriter {
                 // print the suggestion
                 buffer.append(row_num, &replace_tabs(line), Style::NoStyle);
 
-                if is_multiline {
-                    for SubstitutionHighlight { start, end } in parts {
-                        buffer.set_style_range(
-                            row_num,
-                            max_line_num_len + 3 + start,
-                            max_line_num_len + 3 + end,
-                            Style::Addition,
-                            true,
-                        );
-                    }
+                // Colorize addition/replacements with green.
+                for &SubstitutionHighlight { start, end } in highlight_parts {
+                    // Account for tabs when highlighting (#87972).
+                    let tabs: usize = line
+                        .chars()
+                        .take(start)
+                        .map(|ch| match ch {
+                            '\t' => 3,
+                            _ => 0,
+                        })
+                        .sum();
+                    buffer.set_style_range(
+                        row_num,
+                        max_line_num_len + 3 + start + tabs,
+                        max_line_num_len + 3 + end + tabs,
```

Comment:
This assumes the addition/replacement doesn't contain tabs itself. It rarely happens, but it can occur:

<img width="630" alt="image" src="https://user-images.githubusercontent.com/783247/130146173-1a0b2913-015f-455e-8522-280bb7a3de9f.png">

Here, the entire closure up to and including the final `}` is supposed to be highlighted, but i put some tabs in the string literal which broke it.

Reply:
Ah, yes. I assumed that we didn't have tabs in the suggestion because I didn't account for cases where snippets from the user include them :-/

Reply:
I suppose all cases where user snippets are included should be replaced by multispan suggestions. (This specific one is already fixed by https://github.com/rust-lang/rust/pull/87958, but I think there's quite a few places left where this happens.)

Reply:
Could you give me the repro case? I think I fixed this, but the suggestions seem to be doing the right thing for me.

Reply:
With #87958 merged, the test case in the screenshot above doesn't trigger the problem anymore. We'll have to find another case where a bunch of user code gets put verbatim in the suggestion. Grepping for `span_to_source` now. ^^

Reply:
Found one:

<img width="594" alt="Screenshot 2021-08-23 at 15 59 26" src="https://user-images.githubusercontent.com/783247/130460354-79af5596-a37f-448f-bb9e-739ec0b97ec6.png">

```rust
mod a {
    pub const A: String = String::new();
}

fn main() {
    a
    ::	A;
    //^ TAB here
}
```

Reply:
The current code handles this correctly:

![](https://user-images.githubusercontent.com/1606434/130473221-363ace02-0dd0-499d-a6dc-704cef7d4922.png)